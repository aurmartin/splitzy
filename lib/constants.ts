export const primaryHue = 200;

export const Colors = {
  primaryDark: `hsl(${primaryHue} 80% 40%)`,
  primary: `hsl(${primaryHue} 40% 50%)`,
  primaryLight: `hsl(${primaryHue} 40% 80%)`,

  secondary: `hsl(0 0% 70%)`,
  secondaryLight: `hsl(0 0% 80%)`,

  red: `hsl(0 50% 40%)`,

  green: `hsl(120 50% 40%)`,

  background: `hsl(0 0% 94%)`,
};
