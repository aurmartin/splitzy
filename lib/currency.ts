import { type Dinero, type Currency } from "dinero.js";

const MINOR_UNIT_FACTOR: Partial<Record<Currency, number>> = {
  EUR: 100,
  USD: 100,
};

const getAmountColor = (amount: Dinero): string => {
  if (amount.getAmount() > 0) return "hsl(120 70% 30%)";
  if (amount.getAmount() < 0) return "hsl(0 70% 30%)";
  return "black";
};

const getAbsoluteAmount = (amount: Dinero): Dinero => {
  return amount.getAmount() < 0 ? amount.multiply(-1) : amount;
};

const getBalanceMessage = (amount: Dinero): string => {
  if (amount.getAmount() > 0) return "On vous doit";
  if (amount.getAmount() < 0) return "Vous devez";
  return "Vous êtes à jour";
};

const getMinorUnitAmount = (amount: number, currency: Currency): number => {
  const factor = MINOR_UNIT_FACTOR[currency];
  if (!factor) {
    throw new Error(`Unsupported currency: ${currency}`);
  }
  return amount * factor;
};

const getMajorUnitAmount = (amount: number, currency: Currency): number => {
  const factor = MINOR_UNIT_FACTOR[currency];
  if (!factor) {
    throw new Error(`Unsupported currency: ${currency}`);
  }
  return amount / factor;
};

export {
  getAmountColor,
  getAbsoluteAmount,
  getBalanceMessage,
  getMinorUnitAmount,
  getMajorUnitAmount,
};
