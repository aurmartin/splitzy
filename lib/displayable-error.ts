export class DisplayableError extends Error {
  title: string;
  message: string;

  constructor(title: string, message: string) {
    super(`${title}: ${message}`);
    this.title = title;
    this.message = message;
  }
}
