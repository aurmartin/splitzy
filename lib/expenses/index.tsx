export * from "./types";
export * from "./hooks";

export { ExpensesRepository } from "./repository";
export { ExpenseService } from "./service";
