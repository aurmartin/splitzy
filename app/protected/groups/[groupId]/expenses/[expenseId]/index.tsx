import { Screen } from "@/components/screen";
import { Text } from "@/components/text";
import { TopBar } from "@/components/top-bar";
import { useSnackBar } from "@/components/snack-bar";
import { useDelExpense, useExpense, getAmountSplit } from "@/lib/expenses";
import { useGroup } from "@/lib/groups";
import { getLocale } from "@/lib/locale";
import { Ionicons } from "@expo/vector-icons";
import { router, useLocalSearchParams } from "expo-router";
import { ActivityIndicator, ScrollView, View } from "react-native";
import { Pressable } from "react-native-gesture-handler";

export default function ExpenseScreen() {
  const { expenseId, groupId }: { expenseId: string; groupId: string } =
    useLocalSearchParams();

  const snackBar = useSnackBar();
  const group = useGroup(groupId);
  const expense = useExpense(expenseId);
  const delExpense = useDelExpense();

  if (!expense || !group) {
    return <ActivityIndicator />;
  }

  const handleDelete = async () => {
    try {
      await delExpense(expenseId);

      router.back();

      snackBar.show("Dépense supprimée", "success");
    } catch (error) {
      console.error("Delete expense error", error);

      snackBar.show("Une erreur est survenue", "error");
    }
  };

  const handleEdit = () => {
    router.replace(`/protected/groups/${groupId}/expenses/${expenseId}/edit`);
  };

  return (
    <Screen>
      <ScrollView>
        <TopBar
          rightActions={[
            <Pressable
              onPress={handleEdit}
              key="edit"
              style={{
                backgroundColor: "white",
                padding: 8,
                borderRadius: 24,
              }}
              android_ripple={{ color: "hsl(0 0% 90%)" }}
            >
              <Ionicons name="pencil" size={24} color="hsl(0 0% 40%)" />
            </Pressable>,

            <Pressable
              onPress={handleDelete}
              key="delete"
              style={{
                backgroundColor: "hsl(348, 45%, 85%)",
                padding: 8,
                borderRadius: 24,
              }}
              android_ripple={{ color: "hsl(348, 40%, 80%)" }}
            >
              <Ionicons name="trash" size={24} color="hsl(348, 40%, 50%)" />
            </Pressable>,
          ]}
        />

        <Text type="headline" style={{ marginBottom: 16 }}>
          {expense.title}{" "}
          {expense.splitExpense.total.setLocale(getLocale()).toFormat()}
        </Text>

        <Text style={{ marginBottom: 16 }}>
          Payé par {expense.payerName} le{" "}
          {new Date(expense.createdAt).toLocaleDateString()}
        </Text>

        {Object.entries(getAmountSplit(expense).amounts).map(
          ([name, amount]) => (
            <View
              key={name}
              style={{
                marginBottom: 8,
                backgroundColor: "white",
                padding: 14,
                elevation: 1,
                borderRadius: 8,
                flexDirection: "row",
                justifyContent: "space-between",
                alignItems: "center",
              }}
            >
              <Text key={name}>{name}</Text>
              <Text>{amount.setLocale(getLocale()).toFormat()}</Text>
            </View>
          ),
        )}

        {expense.receipt && (
          <View>
            <Text style={{ marginBottom: 8, marginTop: 8 }}>Receipt</Text>

            <View
              style={{
                flex: 1,
                gap: 16,
                backgroundColor: "white",
                padding: 16,
                elevation: 1,
                borderRadius: 8,
              }}
            >
              {expense.receipt.items.map((item) => (
                <View
                  key={item.description}
                  style={{
                    flexDirection: "row",
                    justifyContent: "space-between",
                    alignItems: "center",
                  }}
                >
                  <View>
                    <Text>{item.humanReadableDescription}</Text>
                    <Text
                      type="bodyMedium"
                      style={{ color: "hsl(0, 0%, 40%)" }}
                    >
                      {item.description}
                    </Text>
                  </View>
                  <Text>{item.price.setLocale(getLocale()).toFormat()}</Text>
                </View>
              ))}
            </View>
          </View>
        )}
      </ScrollView>
    </Screen>
  );
}
