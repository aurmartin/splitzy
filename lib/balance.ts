import dinero, { type Dinero } from "dinero.js";
import { type Expense, ExpenseService, useExpenses } from "@/lib/expenses";
import { type Group } from "@/lib/groups";

export type Balance = Record<string, Dinero>;

export type Reimbursement = {
  from: string;
  to: string;
  amount: Dinero;
};

export function computeBalance(group: Group, expenses: Expense[]): Balance {
  const balance = group.members.reduce((acc, member) => {
    acc[member] = dinero({ amount: 0, currency: group.currency });
    return acc;
  }, {} as Balance);

  expenses.forEach((expense) => {
    const payer = expense.payerName;
    const total = expense.splitExpense.total;

    balance[payer] = balance[payer].add(total);

    const amounts = ExpenseService.getAmounts(expense);

    Object.entries(amounts).forEach(([name, amount]) => {
      balance[name] = balance[name].subtract(amount);
    });
  });

  return balance;
}

export const useBalance = (group: Group): Balance => {
  const expenses = useExpenses(group.id);
  return computeBalance(group, expenses);
};

export function computeOptimalReimbursements(
  balance: Balance,
): Reimbursement[] {
  const reimbursements: Reimbursement[] = [];
  const members = Object.keys(balance);

  // Create a copy of the balance object
  const workingBalance = members.reduce((acc, member) => {
    acc[member] = balance[member];
    return acc;
  }, {} as Balance);

  // Continue while there are non-zero balances
  while (members.some((member) => !workingBalance[member].isZero())) {
    // Find the person who owes the most (most negative balance)
    const maxDebtor = members.reduce((a, b) =>
      workingBalance[a].lessThan(workingBalance[b]) ? a : b,
    );

    // Find the person who is owed the most (most positive balance)
    const maxCreditor = members.reduce((a, b) =>
      workingBalance[a].greaterThan(workingBalance[b]) ? a : b,
    );

    // If both balances are effectively zero (less than 1 cent), break
    if (
      workingBalance[maxDebtor].isZero() &&
      workingBalance[maxCreditor].isZero()
    ) {
      break;
    }

    // Calculate the reimbursement amount (minimum of absolute values)
    const debtorAmount = workingBalance[maxDebtor].multiply(-1); // Make positive
    const creditorAmount = workingBalance[maxCreditor];
    const amount = debtorAmount.lessThan(creditorAmount)
      ? debtorAmount
      : creditorAmount;

    // Update the balances
    workingBalance[maxDebtor] = workingBalance[maxDebtor].add(amount);
    workingBalance[maxCreditor] = workingBalance[maxCreditor].subtract(amount);

    // Record the reimbursement
    reimbursements.push({
      from: maxDebtor,
      to: maxCreditor,
      amount,
    });
  }

  return reimbursements;
}
