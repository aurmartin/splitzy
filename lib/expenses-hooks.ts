import { useSystem } from "@/components/system-provider";
import { Expense } from "@/lib/expenses";
import { ExpensesRepository } from "@/lib/expenses-repository";
import { useEffect, useState, useMemo } from "react";

function useExpenses(groupId: string): Expense[] {
  const system = useSystem();
  const initialExpenses = useMemo(
    () => ExpensesRepository.getGroupExpenses(system, groupId),
    [system, groupId],
  );
  const [expenses, setExpenses] = useState<Expense[]>(initialExpenses);

  useEffect(() => {
    return ExpensesRepository.watchGroupExpenses(system, groupId, setExpenses);
  }, [system, groupId]);

  return expenses;
}

function useExpense(id: string): Expense | null {
  const system = useSystem();
  const initialExpense = useMemo(
    () => ExpensesRepository.getExpense(system, id),
    [system, id],
  );
  const [expense, setExpense] = useState<Expense | null>(initialExpense);

  useEffect(() => {
    return ExpensesRepository.watchExpense(system, id, setExpense);
  }, [system, id]);

  return expense;
}

export { useExpense, useExpenses };
