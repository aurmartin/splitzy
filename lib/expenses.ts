import { convertSplit } from "@/components/split-input";
import { useSystem } from "@/components/system-provider";
import { ValidationError } from "@/lib/validation-error";
import { ExpensesRepository } from "@/lib/expenses-repository";
import { System } from "@/lib/system";
import { generateId } from "@/lib/utils";
import { Currency, Dinero } from "dinero.js";
import { useCallback } from "react";

interface BaseSplit {
  type: "percentage" | "amount" | "equal" | "receipt";
  total: Dinero;
  members: string[];
}

interface PercentageSplit extends BaseSplit {
  type: "percentage";
  ratios: { [member: string]: number };
}

interface EqualSplit extends BaseSplit {
  type: "equal";
}

interface AmountSplit extends BaseSplit {
  type: "amount";
  amounts: { [member: string]: Dinero };
}

interface ReceiptSplit extends BaseSplit {
  type: "receipt";
  receipt: Receipt;
}

type Split = PercentageSplit | EqualSplit | AmountSplit | ReceiptSplit;

interface Receipt {
  date: Date;
  total: Dinero;
  title: string;
  currency: Currency;

  items: {
    description: string;
    humanReadableDescription: string;
    price: Dinero;
    paid_for: string[];
  }[];
}

interface Expense {
  id: string;
  groupId: string;
  title: string;
  payerName: string;
  splitExpense: Split;

  createdAt: Date;
  updatedAt: Date;
  deletedAt: Date | undefined;

  receipt: Receipt | null;
}

interface ExpenseCreateParams {
  groupId: Expense["groupId"];
  title: Expense["title"];
  payerName: Expense["payerName"];
  splitExpense: Expense["splitExpense"];
  receipt: Expense["receipt"];
}

interface ExpenseUpdateParams {
  title: Expense["title"];
  payerName: Expense["payerName"];
  splitExpense: Expense["splitExpense"];
  receipt: Expense["receipt"];
}

const useAddExpense = () => {
  const system = useSystem();

  return useCallback(
    (params: ExpenseCreateParams) => addExpense(system, params),
    [system],
  );
};

const useEditExpense = () => {
  const system = useSystem();
  return useCallback(
    (id: string, params: ExpenseUpdateParams) =>
      editExpense(system, id, params),
    [system],
  );
};

const useDelExpense = () => {
  const system = useSystem();
  return useCallback(
    (id: string) => ExpensesRepository.deleteExpense(system, id),
    [system],
  );
};

const addExpense = async (system: System, params: ExpenseCreateParams) => {
  const validParams = validateCreateExpenseParams(params);
  const expense = createExpense(validParams);
  await ExpensesRepository.insertExpense(system, expense);
};

const editExpense = async (
  system: System,
  id: string,
  params: ExpenseUpdateParams,
) => {
  const expense = ExpensesRepository.getExpense(system, id);
  if (expense === null) {
    throw new Error("Expense not found");
  }

  const validParams = validateUpdateExpenseParams(params);
  const updatedExpense = modifyExpense(expense, validParams);
  await ExpensesRepository.updateExpense(system, updatedExpense);
};

const createExpense = (validParams: ExpenseCreateParams): Expense => {
  return {
    ...validParams,
    id: generateId(),
    createdAt: new Date(),
    updatedAt: new Date(),
    deletedAt: undefined,
  };
};

const modifyExpense = (
  expense: Expense,
  params: ExpenseUpdateParams,
): Expense => {
  return {
    ...expense,
    ...params,
    updatedAt: new Date(),
  };
};

const validateCreateExpenseParams = (params: ExpenseCreateParams) => {
  const errors: Record<string, string> = {};

  if (params.title.length === 0) {
    errors.title = "Veuillez entrer un titre";
  }

  if (params.payerName.length === 0) {
    errors.payerName = "Veuillez sélectionner un membre";
  }

  if (Object.keys(errors).length > 0) {
    throw new ValidationError(errors);
  }

  return params;
};

const validateUpdateExpenseParams = (params: ExpenseUpdateParams) => {
  const errors: Record<string, string> = {};

  if (params.title.length === 0) {
    errors.title = "Veuillez entrer un titre";
  }

  if (params.payerName.length === 0) {
    errors.payerName = "Veuillez sélectionner un membre";
  }

  return params;
};

const getAmountSplit = (expense: Expense): AmountSplit => {
  return convertSplit(expense.splitExpense);
};

const useExpense = ExpensesRepository.useExpense;
const useExpenses = ExpensesRepository.useExpenses;

export type {
  AmountSplit,
  EqualSplit,
  Expense,
  PercentageSplit,
  Receipt,
  ReceiptSplit,
  Split,
};

export {
  addExpense,
  getAmountSplit,
  useAddExpense,
  useDelExpense,
  useEditExpense,
  useExpense,
  useExpenses,
};
