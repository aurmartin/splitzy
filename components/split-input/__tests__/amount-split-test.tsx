import {
  AmountSplitInput,
  createAmountSplit,
} from "@/components/split-input/amount-split";
import { type AmountSplit } from "@/lib/expenses";
import { render } from "@testing-library/react-native";
import dinero from "dinero.js";
import * as React from "react";

describe("AmountSplit", () => {
  describe("AmountSplitInput", () => {
    it("renders correctly", () => {
      const mockSplit = {
        type: "amount" as const,
        amounts: { user1: dinero({ amount: 1000, currency: "USD" }) },
        total: dinero({ amount: 1000, currency: "USD" }),
        members: ["user1"],
      } as AmountSplit;

      const tree = render(<AmountSplitInput value={mockSplit} />);
      expect(tree).toMatchSnapshot();
    });
  });

  describe("createAmountSplit", () => {
    it("creates equal splits correctly", () => {
      const total = dinero({ amount: 2000, currency: "USD" });
      const members = ["user1", "user2"];

      const result = createAmountSplit(total, members);

      expect(result.type).toBe("amount");
      expect(Object.keys(result.amounts)).toHaveLength(2);
      expect(result.amounts["user1"].getAmount()).toBe(1000);
      expect(result.amounts["user2"].getAmount()).toBe(1000);
    });

    it("handles no members", () => {
      const total = dinero({ amount: 1000, currency: "USD" });
      const members = [] as string[];

      const result = createAmountSplit(total, members);

      expect(result.type).toBe("amount");
      expect(Object.keys(result.amounts)).toHaveLength(0);
    });
  });
});
