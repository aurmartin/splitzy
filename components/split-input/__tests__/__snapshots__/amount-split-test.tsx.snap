// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmountSplit AmountSplitInput renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "white",
      "borderRadius": 4,
      "flex": 1,
      "gap": 8,
      "padding": 12,
    }
  }
>
  <Text
    style={
      [
        {
          "fontSize": 16,
          "fontWeight": "400",
          "lineHeight": 24,
        },
        {
          "color": "hsl(0, 100%, 30%)",
        },
      ]
    }
  >
    Not implemented
  </Text>
</View>
`;
